L1Trigger/L1THGCalUtilities/plugins/CaloTruthCellsProducer.cc

```diff
@@ -0,0 +1,359 @@
+  auto const& caloParticles(*caloParticlesHandle);
+
+  edm::Handle<l1t::HGCalTriggerCellBxCollection> triggerCellsHandle;
+  event.getByToken(triggerCellsToken_, triggerCellsHandle);
+  auto const& triggerCells(*triggerCellsHandle);
+
+  dummyClustering_.eventSetup(setup);
+  showerShape_.eventSetup(setup);
+
+  edm::ESHandle<HGCalTriggerGeometryBase> geometryHandle;
+  setup.get<CaloGeometryRecord>().get(geometryHandle);
+  auto const& geometry(*geometryHandle);
+
+  std::unordered_map<uint32_t, CaloParticleRef> tcToCalo;
+
+  std::unordered_map<uint32_t, double> hitToEnergy(makeHitMap(event, setup)); // cellId -> sim energy
+  std::unordered_map<uint32_t, std::pair<double, double>> tcToEnergies; // tcId -> {total sim energy, fractioned sim energy}
+
+  for (auto& he : hitToEnergy) {
+    DetId hitId(he.first);
+    uint32_t tcId;
+    try {
+      tcId = geometry.getTriggerCellFromCell(hitId);
+    }
+    catch (cms::Exception const& ex) {
+      edm::LogError("CaloTruthCellsProducer") << ex.what();
+      continue;
+    }
+
+    tcToEnergies[tcId].first += he.second;
+  }
+
+  // used later to order multiclusters
+  std::map<int, CaloParticleRef> orderedCaloRefs;
+
+  for (unsigned iP(0); iP != caloParticles.size(); ++iP) {
+    auto const& caloParticle(caloParticles.at(iP));
+    if (caloParticle.g4Tracks().at(0).eventId().event() != 0) // pileup
+      continue;
+
+    CaloParticleRef ref(caloParticlesHandle, iP);
+
+    SimClusterRefVector const& simClusters(caloParticle.simClusters());
+    for (auto const& simCluster : simClusters) {
+      for (auto const& hAndF : simCluster->hits_and_fractions()) {
+        DetId hitId(hAndF.first);
+        uint32_t tcId;
+        try {
+          tcId = geometry.getTriggerCellFromCell(hitId);
+        }
+        catch (cms::Exception const& ex) {
+          edm::LogError("CaloTruthCellsProducer") << ex.what();
+          continue;
+        }
+
+        tcToCalo.emplace(tcId, ref);
+        tcToEnergies[tcId].second += hitToEnergy[hAndF.first] * hAndF.second;
+      }
+    }
+
+    // ordered by the gen particle index
+    int genIndex(caloParticle.g4Tracks().at(0).genpartIndex() - 1);
+    if (genIndex >= 0) // < 0 shouldn't happen
+      orderedCaloRefs[genIndex] = ref;
+  }
+
+  auto outMap(std::make_unique<CaloToCellsMap>(caloParticlesHandle, triggerCellsHandle));
+  std::unique_ptr<l1t::HGCalTriggerCellBxCollection> outCollection;
+  if (makeCellsCollection_)
+    outCollection = std::move(std::make_unique<l1t::HGCalTriggerCellBxCollection>());
+
+  typedef edm::Ptr<l1t::HGCalTriggerCell> TriggerCellPtr;
+  typedef edm::Ptr<l1t::HGCalCluster> ClusterPtr;
+
+  // ClusteringDummyImpl only considers BX 0, so we dump all cells to one vector
+  std::vector<TriggerCellPtr> triggerCellPtrs;
+
+  // loop through all bunch crossings
+  for (int bx(triggerCells.getFirstBX()); bx <= triggerCells.getLastBX(); ++bx) {
+    for (auto&& cItr(triggerCells.begin(bx)); cItr != triggerCells.end(bx); ++cItr) {
```

Comment:
I don't think a `rvalue` reference is useful here. A standard reference should do the job.

Reply:
begin() returns an iterator by value, so I will just make this `auto`. I like to use `auto&&` when I'm not sure if the function returns by reference or by value (If a reference is returned, `auto&&` becomes a normal lvalue reference). `auto` will cause an unnecessary copy if the reference is returned. But a quick googling says this isn't standard practice, so I won't do it any more.